plugin/firewall/firewall.go

```diff
@@ -0,0 +1,126 @@
+// Package firewall enables filtering on query and response using direct expression as policy.
+// it allows interact with other Policy Engines if those are plugin implementing the Engineer interface
+package firewall
+
+import (
+	"context"
+	"errors"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/firewall/policy"
+	"github.com/coredns/coredns/plugin/firewall/rule"
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/plugin/pkg/nonwriter"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+var (
+	errInvalidAction = errors.New("invalid action")
+)
+
+// ExpressionEngineName is the name associated with built-in rules of Expression type.
+const ExpressionEngineName = "--default--"
+
+// firewall represents a plugin instance that can validate DNS
+// requests and replies using rulelists on the query and/or on the reply
+type firewall struct {
+	engines map[string]policy.Engine
+	query   *rule.List
+	reply   *rule.List
+
+	next plugin.Handler
+}
+
+//New build a new firewall plugin
+func New() (*firewall, error) {
+	pol := &firewall{engines: map[string]policy.Engine{"--default--": policy.NewExprEngine()}}
+	var err error
+	if pol.query, err = rule.NewList(policy.TypeBlock, false); err != nil {
+		return nil, err
+	}
+	if pol.reply, err = rule.NewList(policy.TypeAllow, true); err != nil {
+		return nil, err
+	}
+	return pol, nil
+}
+
+// ServeDNS implements the Handler interface.
+func (p *firewall) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	var (
+		status    = -1
+		respMsg   *dns.Msg
+		err       error
+		queryData = make(map[string]interface{}, 0)
+	)
+
+	state := request.Request{W: w, Req: r}
+
+	// ask policy for the Query Rulelist
+	action, err := p.query.Evaluate(ctx, state, queryData, p.engines)
```

Comment:
what does `queryData` do here? Why is it empty?

Reply:
queryData is a map we need to keep for the response filter evaluation. Therefore it must be at ServeDNS level. But it is really filled when needed by the rule.list.Evaluate function.

Reply:
querydata should not be a map to an empty interface, but use something typed. Make no sense to make it an `interface{}`.

Also: while looking to the `qrdata` plugin, *what* data isn't know at this point?
Also: this argument is  passed by pointer, instead of this method just returning it? It should just return it. Also very curious to learn what *extra* data will put in it that we don't already know